### ISSUE TITLE
findOrCreate creating duplicate Objects (using transaction and hooks)

### ISSUE THREAD

**Issue by xitij** [sequelize/sequelize#9245]
## What are you doing?
I'm using sequelize to manage a db of sporting events. Multiple `Game`s have the same `Venue`. Upon attempting to `findOrCreate()` a venue for 2 different games (in close succession) the venue is getting created twice in the db.

This is the function that is used to `findOrCreate()` the `Venue`
```js
function findOrCreateVenue(v) {
  const where = { name: v.name, city: v.city };
  return sequelize.transaction({
    isolationLevel: sequelize.Transaction.ISOLATION_LEVELS.READ_UNCOMMITTED
  }, (t) => {
    return Venue.findOrCreate({where, defaults: v, transaction: t})
      .spread((foundVenue, created) => {
        if (created) {
          return foundVenue;
        }
        return foundVenue.update(v);
      });
  });
}
```
This is the bit of code from `Venue`.
```js
module.exports = (sequelize, DataTypes) => {
  const Venue = sequelize.define('Venue', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      unique: true,
      allowNull: false,
      primaryKey: true
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        notEmpty: true
      }
    },
    city: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        notEmpty: true
      }
    },
    ...
  }, {
    hooks: {
      afterCreate
    }
  });

  function afterCreate(venue, options) {
    const Topic = sequelize.models.Topic;
    const venueTopic = {
      type: 'venue',
      id: venue.id,
      name: venue.name,
      isDiscoverable: false
    };
    return Topic.create(venueTopic, { transaction: options.transaction });
  }
...
  return Venue;
}
```
This is a minimal snippet from `Topic`
```js
module.exports = (sequelize, DataTypes) => {
  const Topic = sequelize.define('Topic', {
    id: {
      type: DataTypes.UUID,
      allowNull: false,
      primaryKey: true
    },
    type: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        isIn: [['team', 'game', 'competition', 'player','venue']]
      }
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false
    },
    ...
  }, {
    name: {
      singular: 'topic',
      plural: 'topics'
    }
  });

  return Topic;
};
```

## What do you expect to happen?
Since I am using a `transaction` with `READ UNCOMMITTED` I expect the second `findOrCreate()` to find the venue and not create a second with the same name and location (I can guarantee from the dataset that there is no overlap here).

## What is actually happening?
The venue is getting created twice. I'm pruned down the logging a bit to show the important parts, mainly the 2 transactions that are creating the venues.

_SQL Output_
The important transaction ids are `cb99c19b-ed99-453a-b71d-4f2506e95b5f` and `bc3bee74-0480-4909-a379-83afc3d1b37a`
```sql
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): START TRANSACTION;
...
Executing (4954c94d-16b9-4d9c-8cf3-2e19d85eff53): START TRANSACTION;
...
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
Executing (4954c94d-16b9-4d9c-8cf3-2e19d85eff53): SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
...
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): SAVEPOINT "cb99c19b-ed99-453a-b71d-4f2506e95b5f-savepoint-1";
...
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): START TRANSACTION;
Executing (4954c94d-16b9-4d9c-8cf3-2e19d85eff53): SAVEPOINT "4954c94d-16b9-4d9c-8cf3-2e19d85eff53-savepoint-1";
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): SELECT "id", "name", "street", "city", "state", "country", "createdAt", "updatedAt" FROM "Venues" AS "Venue" WHERE "Venue"."name" = 'Holkar Cricket Stadium' AND "Venue"."city" = 'Indore' LIMIT 1;
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
...
Executing (4954c94d-16b9-4d9c-8cf3-2e19d85eff53): SELECT "id", "name", "street", "city", "state", "country", "createdAt", "updatedAt" FROM "Venues" AS "Venue" WHERE "Venue"."name" = 'MA Chidambaram Stadium' AND "Venue"."city" = 'Chennai' LIMIT 1;
...
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): SAVEPOINT "bc3bee74-0480-4909-a379-83afc3d1b37a-savepoint-1";
...
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): CREATE OR REPLACE FUNCTION pg_temp.testfunc(OUT response "Venues", OUT sequelize_caught_exception text) RETURNS RECORD AS $func_91f83b38081048119a83d37a3d84f92a$ BEGIN INSERT INTO "Venues" ("id","name","city","country","createdAt","updatedAt") VALUES ('ed6b12ac-85e9-46fb-a0e9-d0bbe1e6a133','Holkar Cricket Stadium','Indore','IND','2018-03-28 00:53:54.210 +00:00','2018-03-28 00:53:54.210 +00:00') RETURNING * INTO response; EXCEPTION WHEN unique_violation THEN GET STACKED DIAGNOSTICS sequelize_caught_exception = PG_EXCEPTION_DETAIL; END $func_91f83b38081048119a83d37a3d84f92a$ LANGUAGE plpgsql; SELECT (testfunc.response).*, testfunc.sequelize_caught_exception FROM pg_temp.testfunc(); DROP FUNCTION IF EXISTS pg_temp.testfunc();
...
Executing (4954c94d-16b9-4d9c-8cf3-2e19d85eff53): COMMIT;
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): SELECT "id", "name", "street", "city", "state", "country", "createdAt", "updatedAt" FROM "Venues" AS "Venue" WHERE "Venue"."name" = 'Holkar Cricket Stadium' AND "Venue"."city" = 'Indore' LIMIT 1;
...
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): INSERT INTO "Topics" ("id","type","name","order","isDiscoverable","keywords","createdAt","updatedAt") VALUES ('ed6b12ac-85e9-46fb-a0e9-d0bbe1e6a133','venue','Holkar Cricket Stadium',0,false,ARRAY[]::VARCHAR(255)[],'2018-03-28 00:53:54.274 +00:00','2018-03-28 00:53:54.274 +00:00') RETURNING *;
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): CREATE OR REPLACE FUNCTION pg_temp.testfunc(OUT response "Venues", OUT sequelize_caught_exception text) RETURNS RECORD AS $func_bdf2b0a18dc748808e2c2b69df42f0e5$ BEGIN INSERT INTO "Venues" ("id","name","city","country","createdAt","updatedAt") VALUES ('7cbbe7b3-b781-4179-823c-da844b46c511','Holkar Cricket Stadium','Indore','IND','2018-03-28 00:53:54.286 +00:00','2018-03-28 00:53:54.286 +00:00') RETURNING * INTO response; EXCEPTION WHEN unique_violation THEN GET STACKED DIAGNOSTICS sequelize_caught_exception = PG_EXCEPTION_DETAIL; END $func_bdf2b0a18dc748808e2c2b69df42f0e5$ LANGUAGE plpgsql; SELECT (testfunc.response).*, testfunc.sequelize_caught_exception FROM pg_temp.testfunc(); DROP FUNCTION IF EXISTS pg_temp.testfunc();
...
Executing (cb99c19b-ed99-453a-b71d-4f2506e95b5f): COMMIT;
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): INSERT INTO "Topics" ("id","type","name","order","isDiscoverable","keywords","createdAt","updatedAt") VALUES ('7cbbe7b3-b781-4179-823c-da844b46c511','venue','Holkar Cricket Stadium',0,false,ARRAY[]::VARCHAR(255)[],'2018-03-28 00:56:36.881 +00:00','2018-03-28 00:56:36.881 +00:00') RETURNING *;
...
Executing (bc3bee74-0480-4909-a379-83afc3d1b37a): COMMIT;
```

__Dialect:__ postgres
__Dialect version:__ 9.5.12
__Sequelize version:__ 3.30.4
__Tested with latest release:__ No

Sorry for the long post I wanted to follow the template closely and provide as much info as possible. Thanks for the great library.

**Comment by sushantdhiman** [comment]
`findOrCreate` first try to create record by combination of `where` / `defaults`, if that results in `UniqueConstraintError` then it find that record as per `where` options.

As per your description I see no unique constraint over `city + name`, so it just create two records because there was no unique violation.

You may try using [findCreateFind](http://docs.sequelizejs.com/class/lib/model.js~Model.html#static-method-findCreateFind) but that still use unique constraint.

In short you are going to need a unique constraint over `city + name` or any other fields participating in `findOrCreate`

PS> we no longer support v3

**Comment by xitij** [comment]
@sushantdhiman Thanks so for the quick reply.
> findOrCreate first try to create record by combination of where / defaults, if that results in UniqueConstraintError then it find that record as per where options.

Interesting, I must not have read the documentation properly. I just assumed it would attempt to find a document with the `where` clause and then create it if nothing was found. 

> In short you are going to need a unique constraint over city + name or any other fields participating in findOrCreate

Got it, thanks. How do I go about creating a unique constraint that combines 2 fields? Is [this](http://docs.sequelizejs.com/manual/tutorial/models-definition.html#indexes) a good example? I just need to call `sequelize.sync()` after the model is created?

> PS> we no longer support v3

Yes, we'll be upgrading shortly.

**Comment by sushantdhiman** [comment]
> Interesting, I must not have read the documentation properly. I just assumed it would attempt to find a document with the where clause and then create it if nothing was found.

Yeah documentation does not explain this well, I will tag this as documentation 

> Got it, thanks. How do I go about creating a unique constraint that combines 2 fields? Is this a good example? I just need to call sequelize.sync() after the model is created?

Right something like that

```js
sequelize.define('user', {}, {
  indexes: [
    {
      unique: true,
      fields: ['email', 'phone']
    }
  ]
})
```
I dont know how you are managing migrations on production (you need to create that index on production server), but for tests / development `sync()` should work nicely

**Comment by xitij** [comment]
Sorry just to be clear using the example you provided my migration should look something like this...
```js
module.exports = {
  up: function (queryInterface, Sequelize) {
    return queryInterface.addIndex('Users', 'email_phone', {
      unique: true,
      fields: ['email', 'phone']
    });
  },

  down: function (queryInterface, Sequelize) {
    return queryInterface.removeIndex('Users', 'email_phone');
  }
};
```

**Comment by sushantdhiman** [comment]
@xitij LGTM!!

**Comment by xitij** [comment]
@sushantdhiman Cool, thanks for all the help.

Should I close this issue?

**Comment by sushantdhiman** [comment]
No, documentation can be improved

If you want to take a stab at it you can add some comments about unique index [here](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L2068-L2083)

**Comment by sushantdhiman** [comment]
@xitij I checked `findOrCreate` once again, indeed there is a `findOne` call before `create` that explains this issue as opposed to my previous statement, although `create` call still depend on unique constraint.

**Comment by summerkiflain** [comment]
So guys, what is the solution here?, if a `findOne` call is going before the `create` call, shouldn't it find the already existing record, instead of creating a new one with the same `where` clause attributes.

I am also seeing multiple records with the same attributes, although these are happening very rarely but its an issue of data consistency for me, I am looking for a foolproof way of preventing this. wouldn't creating a unique index just fail the findOrCreate command in the create block if find is failing to locate the existing records?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Reading the issue, this looks like a documentation-clarity gap rather than a code defect: the current page doesn't give readers enough to self-diagnose, which is why it surfaces here as an issue instead of being answered by the docs.

Plan: locate the relevant section, confirm what the code actually does today (so the docs match behavior, not intent), tighten the wording, and scan neighboring pages for the same inconsistency — otherwise the fix just relocates the confusion.

One thing to confirm before writing: minimal correction scoped to this page, or also cross-link from related guides if the same gap shows up there?

(claiming via @LeonxLJX)
